src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,272 @@
+    rw [adic_completion.is_integer, adic_completion.is_integer, valuation.map_neg], },
+  simpa only [h] using hx,
+end
+
+/-- Negation on the finite adèle ring. -/
+def neg' (x : finite_adele_ring R K) : finite_adele_ring R K := ⟨-x.val, restr_neg R K x⟩
+
+/-- The product of two finite adèles is a finite adèle. -/
+lemma restr_mul (x y : finite_adele_ring R K) : ∀ᶠ (v : height_one_spectrum R) in filter.cofinite,
+  ((x.val * y.val) v ∈ v.adic_completion_integers K) :=
+begin
+  cases x with x hx,
+  cases y with y hy,
+  rw filter.eventually_cofinite at hx hy ⊢,
+  have h_subset : {v : height_one_spectrum R | ¬ (x * y) v ∈  (v.adic_completion_integers K)} ⊆
+    {v : height_one_spectrum R | ¬ x v ∈ (v.adic_completion_integers K)} ∪
+    {v : height_one_spectrum R | ¬ y v ∈ (v.adic_completion_integers K)},
+  { intros v hv,
+    rw [mem_union, mem_set_of_eq, mem_set_of_eq],
+    rw mem_set_of_eq at hv,
+    contrapose! hv,
+    rw [adic_completion.is_integer, adic_completion.is_integer] at hv,
+    have h_mul : valued.v (x v * y v) = (valued.v (x v)) * (valued.v (y v))
+    := (valued.v).map_mul' (x v) (y v),
+    rw [adic_completion.is_integer, pi.mul_apply, h_mul],
+    exact @mul_le_one' (with_zero (multiplicative ℤ)) _ _
+      (ordered_comm_monoid.to_covariant_class_left _) _ _ hv.left hv.right  },
+  exact (hx.union hy).subset h_subset,
+end
+
+/-- Multiplication on the finite adèle ring. -/
+def mul' (x y : finite_adele_ring R K) : finite_adele_ring R K :=
+⟨x.val * y.val, restr_mul R K x y⟩
+
+/-- The tuple `(1)_v` is a finite adèle. -/
+lemma restr_one : ∀ᶠ (v : height_one_spectrum R) in filter.cofinite,
+  ((1 : v.adic_completion K) ∈ v.adic_completion_integers K) :=
+begin
+  rw filter.eventually_cofinite,
+  have h_empty : {v : height_one_spectrum R |
+    ¬ ((1 : v.adic_completion K) ∈ v.adic_completion_integers K)} = ∅,
+  { ext v, rw [mem_empty_iff_false, iff_false], intro hv,
+    rw mem_set_of_eq at hv, apply hv, rw adic_completion.is_integer,
+    exact le_of_eq valued.v.map_one' },
+  rw h_empty,
+  exact finite_empty,
+end
+
+/-- `finite_adele_ring R K` is a commutative additive group. -/
+instance : add_comm_group (finite_adele_ring R K) :=
+{ add          := add' R K,
+  add_assoc    := λ ⟨x, hx⟩ ⟨y, hy⟩ ⟨z, hz⟩,
+  by { dsimp only [add'], rw [subtype.mk_eq_mk], exact add_assoc _ _ _, },
+  zero         := ⟨0, restr_zero R K⟩,
+  zero_add     := λ x, by { simp_rw [add', zero_add, subtype.val_eq_coe, subtype.coe_eta] },
+  add_zero     := λ x, by { simp_rw [add', add_zero, subtype.val_eq_coe, subtype.coe_eta] },
+  neg          := λ x, ⟨-x.val, restr_neg R K x⟩,
+  add_left_neg := λ x, by { unfold_projs,  dsimp only [add'], ext,
+    rw [subtype.coe_mk, subtype.coe_mk, pi.add_apply, add_left_neg], refl, },
+  add_comm     := λ x y, by { unfold_projs,  dsimp only [add'], ext,
+    rw [subtype.coe_mk, subtype.coe_mk, pi.add_apply, pi.add_apply, add_comm], }}
+
+/-- `finite_adele_ring R K` is a commutative ring. -/
```

Comment:
Can you also add the algebra structure? It shouldn't be a lot of work (and maybe also `algebra (finite_adele_ring R K) K_hat` and friends, with the relevant `is_scalar_tower` instance).

Reply:
This would take a bit of work since it requires to prove some `restr_smul ` lemma. Would it be ok to leave this for a future PR?

Reply:
OK!